File: app/lib/models/game.dart
class Game {
  final String id;
  final String organiserId;
  final String venueId;

  /// Set when this game was materialised from a recurring series; null for
  /// one-offs.
  final String? seriesId;

  final String title;
  final String? description;
  final DateTime startsAt;
  final int durationMinutes;
  final int maxPlayers;
  final int currentPlayers;
  final int skillLevel;
  final String status;
  final bool isRecurring;
  final bool isPrivate;

  /// `squad_only` games are reserved for the organiser's squad and hidden from
  /// marketplace discovery; `public` games are open to anyone.
  final String visibility;

  /// When the game reached the marketplace, whether opened by the organiser or
  /// automatically because it was still short close to kick-off.
  final DateTime? marketplaceOpenedAt;

  /// Set once the "still needs players" push has gone out.
  final DateTime? marketplaceNotifiedAt;

  final int? costPerPlayer;
  final String? currency;

  const Game({
    required this.id,
    required this.organiserId,
    required this.venueId,
    this.seriesId,
    required this.title,
    this.description,
    required this.startsAt,
    this.durationMinutes = 60,
    this.maxPlayers = 10,
    this.currentPlayers = 0,
    this.skillLevel = 3,
    this.status = 'open',
    this.isRecurring = false,
    this.isPrivate = false,
    this.visibility = 'public',
    this.marketplaceOpenedAt,
    this.marketplaceNotifiedAt,
    this.costPerPlayer,
    this.currency,
  });

  bool get isSquadOnly => visibility == 'squad_only';
  bool get isFull => currentPlayers >= maxPlayers;
  int get slotsLeft => (maxPlayers - currentPlayers).clamp(0, maxPlayers);

  factory Game.fromJson(Map<String, dynamic> json) {
    return Game(
      id: json['id'] as String,
      organiserId: json['organiser_id'] as String,
      venueId: json['venue_id'] as String,
      seriesId: json['series_id'] as String?,
      title: json['title'] as String,
      description: json['description'] as String?,
      startsAt: DateTime.parse(json['starts_at'] as String),
      durationMinutes: json['duration_minutes'] as int? ?? 60,
      maxPlayers: json['max_players'] as int? ?? 10,
      currentPlayers: json['current_players'] as int? ?? 0,
      skillLevel: json['skill_level'] as int? ?? 3,
      status: json['status'] as String? ?? 'open',
      isRecurring: json['is_recurring'] as bool? ?? false,
      isPrivate: json['is_private'] as bool? ?? false,
      visibility: json['visibility'] as String? ?? 'public',
      marketplaceOpenedAt: json['marketplace_opened_at'] == null
          ? null
          : DateTime.parse(json['marketplace_opened_at'] as String),
      marketplaceNotifiedAt: json['marketplace_notified_at'] == null
          ? null
          : DateTime.parse(json['marketplace_notified_at'] as String),
      costPerPlayer: json['cost_per_player'] as int?,
      currency: json['currency'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'organiser_id': organiserId,
        'venue_id': venueId,
        'series_id': seriesId,
        'title': title,
        'description': description,
        'starts_at': startsAt.toIso8601String(),
        'duration_minutes': durationMinutes,
        'max_players': maxPlayers,
        'current_players': currentPlayers,
        'skill_level': skillLevel,
        'status': status,
        'is_recurring': isRecurring,
        'is_private': isPrivate,
        'visibility': visibility,
        'marketplace_opened_at': marketplaceOpenedAt?.toIso8601String(),
        'marketplace_notified_at': marketplaceNotifiedAt?.toIso8601String(),
        'cost_per_player': costPerPlayer,
        'currency': currency,
      };
}

File: api/models/game.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, func
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column

from database import Base


class Game(Base):
    __tablename__ = "games"

    id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    organiser_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    venue_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("venues.id"), nullable=False
    )
    # Set when this game was materialised from a recurring series. NULL for
    # one-off games, so a game stays the bookable unit either way.
    series_id: Mapped[uuid.UUID | None] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("game_series.id"), nullable=True
    )
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(String(1000))
    starts_at: Mapped[str] = mapped_column(DateTime(timezone=True), nullable=False)
    duration_minutes: Mapped[int] = mapped_column(Integer, default=60)
    max_players: Mapped[int] = mapped_column(Integer, default=10)
    current_players: Mapped[int] = mapped_column(Integer, default=0)
    skill_level: Mapped[int] = mapped_column(Integer, default=3)
    status: Mapped[str] = mapped_column(String(20), default="open")  # open, full, cancelled, completed
    is_recurring: Mapped[bool] = mapped_column(Boolean, default=False)
    is_private: Mapped[bool] = mapped_column(Boolean, default=False)

    # Marketplace visibility. `squad_only` games are hidden from public
    # discovery and reserved for the organiser's squad; `public` games appear
    # in the marketplace for anyone to claim. Defaults to `public` so one-off
    # games behave exactly as they did before this existed.
    visibility: Mapped[str] = mapped_column(
        String(20), default="public", nullable=False
    )  # squad_only, public
    # When the game reached the marketplace, whether opened by the organiser or
    # automatically because it was still short close to kick-off.
    marketplace_opened_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    # Set once the "still short" push has gone out, so it only ever fires once.
    marketplace_notified_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    # Phase 2: payment fields (nullable until Stripe integration)
    cost_per_player: Mapped[int | None] = mapped_column(Integer, nullable=True)
    currency: Mapped[str | None] = mapped_column(String(3), nullable=True)
    created_at: Mapped[str] = mapped_column(server_default=func.now())
    updated_at: Mapped[str] = mapped_column(server_default=func.now(), onupdate=func.now())
